### PR TITLE
Fix margin list for xmonad layouts

### PR DESCRIPTION
**Pull request by ylpoonlg**
Add support for margin list for xmonad layouts as previously ignored in [#2041](https://github.com/qtile/qtile/pull/2041).
The middle margin is calculated as the maximum of the left and right (or top and bottom) margin so that it is not double counted.
I'm not sure if this should be the expected behavior but I think it makes the most sense to make it the union of two adjacent margins.

### CODE REVIEW

**Comment by elParaguayo** [comment]
According to https://github.com/qtile/qtile/issues/3337, there is no double counting of margins in the xmonad layouts.

I'll look at your code properly later on but I've started the tests in the meantime.

**Comment by elParaguayo** [comment]
Re double counting: if a user sets a margin of `5`, I would not expect to have a gap of `10` between windows.

**Comment by ylpoonlg** [comment]
Sorry for the unclear context, but what I mean is that the xmonad layouts were ignored because it didn't have a strategy to combine the left and right margin into one. This PR will not break any configs, it will still work the same way when given an integer.
Example: if margin = [10, 20, 30, 40], the middle margin will be 40 (max of 20 and 40), and the horizontal margins on the secondary pane will be 30 (max of 10 and 30).

File: libqtile/layout/xmonad.py [hunk 1]
```diff
@@ -364,8 +369,23 @@ def _configure_specific(self, client, screen_rect, px, cidx):
...
+                margin[2],
+                max(margin[1], margin[3]),
+            ]
+            margin_right = [
+                margin[0],
+                max(margin[1], margin[3]),
+                margin[2],
+                margin[3],
+            ]
+
```
Comment (elParaguayo) [inline libqtile/layout/xmonad.py:388]:
```suggestion
            # margin list
            if self.align == self._left:
                margin = [
                    margin[0],
                    margin[1],
                    margin[2],
                    max(margin[1], margin[3]),
                ]
            else:
                margin = [
                    margin[0],
                    max(margin[1], margin[3]),
                    margin[2],
                    margin[3],
                ]

```

File: libqtile/layout/xmonad.py [hunk 2]
```diff
@@ -374,23 +394,33 @@ def _configure_specific(self, client, screen_rect, px, cidx):
                 height - 2 * self.border_width,
                 self.border_width,
                 px,
-                margin=self.margin,
+                margin=margin_left if self.align == self._left else margin_right,
```
Comment (elParaguayo) [inline libqtile/layout/xmonad.py:397]:
```suggestion
                margin=margin,
```

File: libqtile/layout/xmonad.py [hunk 3]
```diff
@@ -374,23 +394,33 @@ def _configure_specific(self, client, screen_rect, px, cidx):
...
+                margin[2] + 2 * self.border_width,
+                margin[3],
+            ]
+            margin_right = [
+                margin[0],
+                margin[1],
+                margin[2] + 2 * self.border_width,
+                2 * self.border_width,
+            ]
+
```
Comment (elParaguayo) [inline libqtile/layout/xmonad.py:415]:
```suggestion

            # margin list
            if self.align == self._left:
                margin = [
                    margin[0],
                    2 * self.border_width,
                    margin[2] + 2 * self.border_width,
                    margin[3],
                ]
            else:
                margin = [
                    margin[0],
                    margin[1],
                    margin[2] + 2 * self.border_width,
                    2 * self.border_width,
                ]

```

File: libqtile/layout/xmonad.py [hunk 4]
```diff
@@ -374,23 +394,33 @@ def _configure_specific(self, client, screen_rect, px, cidx):
...
                 self.screen_rect.height,
                 self.border_width,
                 px,
-                margin=[
-                    self.margin,
-                    2 * self.border_width,
-                    self.margin + 2 * self.border_width,
-                    self.margin,
-                ],
+                margin=margin_left if self.align == self._left else margin_right,
```
Comment (elParaguayo) [inline libqtile/layout/xmonad.py:423]:
```suggestion
                margin=margin,
```

File: libqtile/layout/xmonad.py [hunk 5]
```diff
@@ -916,8 +951,23 @@ def _configure_specific(self, client, screen_rect, px, cidx):
...
+                margin[2],
+                margin[3],
+            ]
+            margin_down = [
+                margin[0],
+                margin[1],
+                max(margin[0], margin[2]),
+                margin[3],
+            ]
+
```
Comment (elParaguayo) [inline libqtile/layout/xmonad.py:970]:
```suggestion
            # margin list
            if self.align == self._up:
                margin = [
                    max(margin[0], margin[2]),
                    margin[1],
                    margin[2],
                    margin[3],
                ]
            else:
                margin = [
                    margin[0],
                    margin[1],
                    max(margin[0], margin[2]),
                    margin[3],
                ]

```

File: libqtile/layout/xmonad.py [hunk 6]
```diff
@@ -926,23 +976,33 @@ def _configure_specific(self, client, screen_rect, px, cidx):
...
+                2 * self.border_width,
+                margin[3],
+            ]
+            margin_down = [
+                2 * self.border_width,
+                margin[1] + 2 * self.border_width,
+                margin[2],
+                margin[3],
+            ]
+
```
Comment (elParaguayo) [inline libqtile/layout/xmonad.py:997]:
```suggestion
            # margin list
            if self.align == self._up:
                margin = [
                    margin[0],
                    margin[1] + 2 * self.border_width,
                    2 * self.border_width,
                    margin[3],
                ]
            else:
                margin = [
                    2 * self.border_width,
                    margin[1] + 2 * self.border_width,
                    margin[2],
                    margin[3],
                ]

```

File: libqtile/layout/xmonad.py [hunk 7]
```diff
@@ -926,23 +976,33 @@ def _configure_specific(self, client, screen_rect, px, cidx):
...
                 height_main,
                 self.border_width,
                 px,
-                margin=[
-                    self.margin,
-                    self.margin + 2 * self.border_width,
-                    2 * self.border_width,
-                    self.margin,
-                ],
+                margin=margin_up if self.align == self._up else margin_down,
```
Comment (elParaguayo) [inline libqtile/layout/xmonad.py:1005]:
```suggestion
                margin=margin,
```

**Review by elParaguayo — CHANGES_REQUESTED** [review]

**Comment by elParaguayo** [comment]
I've made some suggestions which would remove some unnecessary calculations.

I haven't tried this code yet though but will run the rest suite to see if the current tests still pass with your changes.

**Comment by elParaguayo** [comment]
@ylpoonlg One question for you on this: if you've changed the margin calculations here why aren't the tests (`test_wide_margins` and `test_tall_margins`) now failing? Is this because they only open 2 windows?

**Comment by elParaguayo** [comment]
OK - are you able to add a test for this or do you need me to do it?

**Comment by ylpoonlg** [comment]
I think I should be able to do it, thanks

**Comment by elParaguayo** [comment]
Ignore the failed test. That's unrelated to your test.

Could you just add some comments to your test to explain why the dimensions/positions are equal to the numbers you're checking.

**Comment by ylpoonlg** [comment]
I have added some comments for one of the cases, not sure if this is enough to make it more understandable. There are too many different cases for the main pane and client panes.

File: test/layouts/test_xmonad.py [hunk 8]
```diff
@@ -255,8 +266,30 @@ def test_tall_margins(manager):
     assert_focused(manager, "one")
     assert_dimensions(manager, 4, 4, 392, 588)
 
+    manager.test_window("three")
+    assert_focused(manager, "three")
+    assert_dimensions(manager, 404, 4, 388, 288)
+
+    manager.c.layout.next()
+    assert_focused(manager, "two")
+    assert_dimensions(manager, 404, 300, 388, 292)
```
Comment (elParaguayo) [inline test/layouts/test_xmonad.py:275]: `two` and `three` have different heights. That doesn't seem right. Is that the issue you were trying to fix?

File: test/layouts/test_xmonad.py [hunk 9]
```diff
@@ -255,8 +266,30 @@ def test_tall_margins(manager):
...
+    # y = screen_height / 2
+    # width = screen_width * 0.5 - margin[1] - margin[3] - 2 * border
+    # height = screen_height / 2 - margin[0] - margin[2] - 2 * border
+    #            + min(margin[0], margin[2]) # Combine top margin
+    assert_focused(manager, "two")
+    assert_dimensions(manager, 408, 300, 384, 290)
+
+    manager.c.layout.previous()
+    assert_focused(manager, "three")
+    assert_dimensions(manager, 408, 2, 384, 288)
```
Comment (elParaguayo) [inline test/layouts/test_xmonad.py:289]: `two` and `three` are still different heights (but are closer now). Is this a rounding issue?

File: test/layouts/test_xmonad.py [hunk 10]
```diff
@@ -255,8 +266,30 @@ def test_tall_margins(manager):
...
+    assert_dimensions(manager, 404, 4, 388, 288)
+
+    manager.c.layout.next()
+    assert_focused(manager, "two")
+    assert_dimensions(manager, 404, 300, 388, 292)
+
+    manager.c.next_layout()
+
+    # x = screen_width * 0.5 + margin[3]
+    # y = screen_height / 2
```
Comment (elParaguayo) [inline test/layouts/test_xmonad.py:280]: Shouldn't there be a an adjustment for the margin here too? Or is the whole margin applied to the window above it?

(Thinking out loud here!) I guess you can't distribute the layout evenly between windows.

File: test/layouts/test_xmonad.py [hunk 11]
```diff
@@ -269,6 +302,29 @@ def test_wide_margins(manager):
...
+    assert_dimensions(manager, 4, 304, 388, 288)
+
+    manager.c.layout.next()
+    assert_focused(manager, "two")
+    assert_dimensions(manager, 400, 304, 392, 288)
+
+    manager.c.next_layout()
+
+    # x = screen_width / 2 + margin[3] - min(margin[1], margin[3])
+    # y = screen_height * 0.5 + max(margin[0], margin[2])
```
Comment (elParaguayo) [inline test/layouts/test_xmonad.py:316]: Isn't this too big? Top window finishes at y=296, you're now at y=306.

**Review by elParaguayo — COMMENTED** [review]
Thanks for updating the test.

I'm still a bit confused by the logic so have added some questions for you.

**Comment by ylpoonlg** [comment]
Sorry, I know it is quite confusing. I actually didn't change how it handles combining the margins. I have just changed the values of the margins. If you take a look at the original source code of the layout, you can see how the dimensions were calculated.

I've put together a simple diagram for a test case:

qtile_xmonad_layout

**Pull request closed without merging**